File: Sources/CococoJinja/Expression/Lexer.swift
//
//  Lexer.swift
//  CococoJinja
//
//  Copyright Triple C Labs GmbH
//
//  Tokenizes the contents of a single `{{ ... }}` expression span. The template
//  layer (splitting text from expression spans) lives in the evaluator; this
//  operates purely on expression source. See docs/workflow-expression-dialect.md §4.
//

import Foundation

enum Lexer {
    /// Tokenize an expression source string into tokens, terminated by `.eof`.
    /// Offsets are measured in Unicode scalars from the start of `source`.
    static func tokenize(_ source: String) throws -> [Token] {
        var scanner = Scanner(source)
        var tokens: [Token] = []
        while let token = try scanner.next() {
            tokens.append(token)
            if token.kind == .eof { break }
        }
        return tokens
    }

    // MARK: - Scanner

    private struct Scanner {
        private let chars: [Character]
        private var pos = 0

        init(_ source: String) {
            self.chars = Array(source)
        }

        private var isAtEnd: Bool { pos >= chars.count }

        private func peek(_ ahead: Int = 0) -> Character? {
            let i = pos + ahead
            return i < chars.count ? chars[i] : nil
        }

        @discardableResult
        private mutating func advance() -> Character {
            defer { pos += 1 }
            return chars[pos]
        }

        private mutating func match(_ c: Character) -> Bool {
            guard peek() == c else { return false }
            pos += 1
            return true
        }

        mutating func next() throws -> Token? {
            skipWhitespace()
            let start = pos
            guard !isAtEnd else { return Token(.eof, at: start) }

            let c = advance()
            switch c {
            case "(": return Token(.lparen, at: start)
            case ")": return Token(.rparen, at: start)
            case "[": return Token(.lbracket, at: start)
            case "]": return Token(.rbracket, at: start)
            case "{": return Token(.lbrace, at: start)
            case "}": return Token(.rbrace, at: start)
            case ",": return Token(.comma, at: start)
            case ":": return Token(.colon, at: start)
            case ".":
                // A leading-dot number like `.5` is not supported; `.` is always
                // member access here.
                return Token(.dot, at: start)
            case "|": return Token(.pipe, at: start)
            case "+": return Token(.plus, at: start)
            case "-": return Token(.minus, at: start)
            case "*": return Token(.star, at: start)
            case "%": return Token(.percent, at: start)
            case "~": return Token(.tilde, at: start)
            case "/":
                return Token(match("/") ? .slashSlash : .slash, at: start)
            case "=":
                guard match("=") else {
                    throw ExpressionError.lex("unexpected '='; did you mean '=='?", at: start)
                }
                return Token(.eq, at: start)
            case "!":
                guard match("=") else {
                    throw ExpressionError.lex("unexpected '!'; use 'not' for negation", at: start)
                }
                return Token(.neq, at: start)
            case "<":
                return Token(match("=") ? .lte : .lt, at: start)
            case ">":
                return Token(match("=") ? .gte : .gt, at: start)
            case "'", "\"":
                return try string(opening: c, start: start)
            default:
                if c.isNumber {
                    return try number(start: start)
                }
                if c.isLetter || c == "_" {
                    return identifierOrKeyword(start: start)
                }
                throw ExpressionError.lex("unexpected character '\(c)'", at: start)
            }
        }

        private mutating func skipWhitespace() {
            while let c = peek(), c == " " || c == "\t" || c == "\n" || c == "\r" {
                pos += 1
            }
        }

        // MARK: literals

        private mutating func number(start: Int) throws -> Token {
            // `start` is the index of the first digit (already consumed).
            var isDouble = false
            while let c = peek(), c.isNumber { pos += 1 }
            // Fractional part — only if a digit follows the dot (so `x.0` member
            // access on a number isn't mis-lexed; here it's a numeric literal).
            if peek() == ".", let after = peek(1), after.isNumber {
                isDouble = true
                pos += 1 // consume '.'
                while let c = peek(), c.isNumber { pos += 1 }
            }
            // Exponent.
            if let c = peek(), c == "e" || c == "E" {
                isDouble = true
                pos += 1
                if let s = peek(), s == "+" || s == "-" { pos += 1 }
                guard let d = peek(), d.isNumber else {
                    throw ExpressionError.lex("malformed exponent in number literal", at: start)
                }
                while let c = peek(), c.isNumber { pos += 1 }
            }
            let text = String(chars[start..<pos])
            if isDouble {
                guard let v = Double(text) else {
                    throw ExpressionError.lex("invalid number literal '\(text)'", at: start)
                }
                return Token(.double(v), at: start)
            }
            guard let v = Int(text) else {
                // Integer literal that overflows Int — fall back to double.
                guard let d = Double(text) else {
                    throw ExpressionError.lex("invalid number literal '\(text)'", at: start)
                }
                return Token(.double(d), at: start)
            }
            return Token(.int(v), at: start)
        }

        private mutating func string(opening quote: Character, start: Int) throws -> Token {
            var value = ""
            while let c = peek() {
                pos += 1
                if c == quote {
                    return Token(.string(value), at: start)
                }
                if c == "\\" {
                    guard let esc = peek() else {
                        throw ExpressionError.lex("unterminated escape in string literal", at: start)
                    }
                    pos += 1
                    switch esc {
                    case "n": value.append("\n")
                    case "t": value.append("\t")
                    case "r": value.append("\r")
                    case "\\": value.append("\\")
                    case "'": value.append("'")
                    case "\"": value.append("\"")
                    case "0": value.append("\0")
                    default:
                        throw ExpressionError.lex("invalid escape '\\\(esc)' in string literal", at: start)
                    }
                } else {
                    value.append(c)
                }
            }
            throw ExpressionError.lex("unterminated string literal", at: start)
        }

        private mutating func identifierOrKeyword(start: Int) -> Token {
            while let c = peek(), c.isLetter || c.isNumber || c == "_" { pos += 1 }
            let text = String(chars[start..<pos])
            switch text {
            case "true": return Token(.kwTrue, at: start)
            case "false": return Token(.kwFalse, at: start)
            case "null", "none": return Token(.kwNull, at: start)
            case "and": return Token(.kwAnd, at: start)
            case "or": return Token(.kwOr, at: start)
            case "not": return Token(.kwNot, at: start)
            case "in": return Token(.kwIn, at: start)
            case "if": return Token(.kwIf, at: start)
            case "else": return Token(.kwElse, at: start)
            default: return Token(.identifier(text), at: start)
            }
        }
    }
}

File: ts/src/lexer.ts
// lexer.ts — port of Expression/Lexer.swift
//
// Tokenizes the contents of a single `{{ ... }}` expression span. Offsets are
// measured in characters (code points) from the start of `source`, matching the
// Swift scanner's `Array(source)` indexing for the ASCII / BMP inputs the
// dialect accepts.

import type { Token, TokenKind } from "./ast.ts";
import { ExpressionError } from "./errors.ts";

// Swift's Character.isNumber / isLetter are Unicode-aware. Identifiers here use
// the same "letter, digit, or underscore" rule; we use Unicode-property regexes
// so non-ASCII letters lex as identifiers too (as they do in Swift).
const RE_DIGIT = /\p{Nd}/u;
const RE_LETTER = /\p{L}/u;

function isNumberChar(c: string): boolean {
	return RE_DIGIT.test(c);
}
function isLetterChar(c: string): boolean {
	return RE_LETTER.test(c);
}
function isIdentContinue(c: string): boolean {
	return isLetterChar(c) || isNumberChar(c) || c === "_";
}

export function tokenize(source: string): Token[] {
	const chars = Array.from(source);
	const tokens: Token[] = [];
	let pos = 0;

	const isAtEnd = () => pos >= chars.length;
	const peek = (ahead = 0): string | undefined => {
		const i = pos + ahead;
		return i < chars.length ? chars[i] : undefined;
	};
	const match = (c: string): boolean => {
		if (peek() === c) {
			pos += 1;
			return true;
		}
		return false;
	};

	const skipWhitespace = () => {
		while (true) {
			const c = peek();
			if (c === " " || c === "\t" || c === "\n" || c === "\r") pos += 1;
			else break;
		}
	};

	const tok = (kind: TokenKind, at: number): Token => ({ kind, offset: at });

	const scanNumber = (start: number): Token => {
		let isDouble = false;
		while (peek() !== undefined && isNumberChar(peek()!)) pos += 1;
		// Fractional part — only if a digit follows the dot.
		if (peek() === "." && peek(1) !== undefined && isNumberChar(peek(1)!)) {
			isDouble = true;
			pos += 1; // consume '.'
			while (peek() !== undefined && isNumberChar(peek()!)) pos += 1;
		}
		// Exponent.
		const e = peek();
		if (e === "e" || e === "E") {
			isDouble = true;
			pos += 1;
			const s = peek();
			if (s === "+" || s === "-") pos += 1;
			const d = peek();
			if (d === undefined || !isNumberChar(d)) {
				throw ExpressionError.lex("malformed exponent in number literal", start);
			}
			while (peek() !== undefined && isNumberChar(peek()!)) pos += 1;
		}
		const text = chars.slice(start, pos).join("");
		if (isDouble) {
			const v = Number(text);
			if (!Number.isFinite(v) && text !== "inf") {
				throw ExpressionError.lex(`invalid number literal '${text}'`, start);
			}
			return tok({ t: "double", value: v }, start);
		}
		// Integer literal. Swift falls back to double when the value overflows Int
		// (64-bit). In JS we fall back once it exceeds the exact-integer range.
		const v = Number(text);
		if (Number.isSafeInteger(v)) {
			return tok({ t: "int", value: v }, start);
		}
		// Overflow → double.
		if (Number.isFinite(v)) {
			return tok({ t: "double", value: v }, start);
		}
		throw ExpressionError.lex(`invalid number literal '${text}'`, start);
	};

	const scanString = (quote: string, start: number): Token => {
		let value = "";
		while (peek() !== undefined) {
			const c = peek()!;
			pos += 1;
			if (c === quote) {
				return tok({ t: "string", value }, start);
			}
			if (c === "\\") {
				const esc = peek();
				if (esc === undefined) {
					throw ExpressionError.lex("unterminated escape in string literal", start);
				}
				pos += 1;
				switch (esc) {
					case "n":
						value += "\n";
						break;
					case "t":
						value += "\t";
						break;
					case "r":
						value += "\r";
						break;
					case "\\":
						value += "\\";
						break;
					case "'":
						value += "'";
						break;
					case '"':
						value += '"';
						break;
					case "0":
						value += "\0";
						break;
					default:
						throw ExpressionError.lex(
							`invalid escape '\\${esc}' in string literal`,
							start,
						);
				}
			} else {
				value += c;
			}
		}
		throw ExpressionError.lex("unterminated string literal", start);
	};

	const scanIdentifierOrKeyword = (start: number): Token => {
		while (peek() !== undefined && isIdentContinue(peek()!)) pos += 1;
		const text = chars.slice(start, pos).join("");
		switch (text) {
			case "true":
				return tok({ t: "kwTrue" }, start);
			case "false":
				return tok({ t: "kwFalse" }, start);
			case "null":
			case "none":
				return tok({ t: "kwNull" }, start);
			case "and":
				return tok({ t: "kwAnd" }, start);
			case "or":
				return tok({ t: "kwOr" }, start);
			case "not":
				return tok({ t: "kwNot" }, start);
			case "in":
				return tok({ t: "kwIn" }, start);
			case "if":
				return tok({ t: "kwIf" }, start);
			case "else":
				return tok({ t: "kwElse" }, start);
			default:
				return tok({ t: "identifier", value: text }, start);
		}
	};

	while (true) {
		skipWhitespace();
		const start = pos;
		if (isAtEnd()) {
			tokens.push(tok({ t: "eof" }, start));
			break;
		}
		const c = chars[pos]!;
		pos += 1;
		switch (c) {
			case "(":
				tokens.push(tok({ t: "lparen" }, start));
				continue;
			case ")":
				tokens.push(tok({ t: "rparen" }, start));
				continue;
			case "[":
				tokens.push(tok({ t: "lbracket" }, start));
				continue;
			case "]":
				tokens.push(tok({ t: "rbracket" }, start));
				continue;
			case "{":
				tokens.push(tok({ t: "lbrace" }, start));
				continue;
			case "}":
				tokens.push(tok({ t: "rbrace" }, start));
				continue;
			case ",":
				tokens.push(tok({ t: "comma" }, start));
				continue;
			case ":":
				tokens.push(tok({ t: "colon" }, start));
				continue;
			case ".":
				tokens.push(tok({ t: "dot" }, start));
				continue;
			case "|":
				tokens.push(tok({ t: "pipe" }, start));
				continue;
			case "+":
				tokens.push(tok({ t: "plus" }, start));
				continue;
			case "-":
				tokens.push(tok({ t: "minus" }, start));
				continue;
			case "*":
				tokens.push(tok({ t: "star" }, start));
				continue;
			case "%":
				tokens.push(tok({ t: "percent" }, start));
				continue;
			case "~":
				tokens.push(tok({ t: "tilde" }, start));
				continue;
			case "/":
				tokens.push(tok({ t: match("/") ? "slashSlash" : "slash" }, start));
				continue;
			case "=":
				if (!match("=")) {
					throw ExpressionError.lex("unexpected '='; did you mean '=='?", start);
				}
				tokens.push(tok({ t: "eq" }, start));
				continue;
			case "!":
				if (!match("=")) {
					throw ExpressionError.lex("unexpected '!'; use 'not' for negation", start);
				}
				tokens.push(tok({ t: "neq" }, start));
				continue;
			case "<":
				tokens.push(tok({ t: match("=") ? "lte" : "lt" }, start));
				continue;
			case ">":
				tokens.push(tok({ t: match("=") ? "gte" : "gt" }, start));
				continue;
			case "'":
			case '"':
				tokens.push(scanString(c, start));
				continue;
			default:
				if (isNumberChar(c)) {
					// Rewind one — scanNumber expects to consume from `start`.
					pos = start;
					tokens.push(scanNumber(start));
					continue;
				}
				if (isLetterChar(c) || c === "_") {
					pos = start;
					tokens.push(scanIdentifierOrKeyword(start));
					continue;
				}
				throw ExpressionError.lex(`unexpected character '${c}'`, start);
		}
	}

	return tokens;
}
